packages/snjs/lib/services/history/history_manager.ts

```diff
@@ -226,24 +317,95 @@ export class SNHistoryManager extends PureService {
       return undefined;
     }
-    const payload = CreateMaxPayloadFromAnyObject(revision as any, {
-      uuid: revision.item_uuid,
-    });
+    const payload = CreateMaxPayloadFromAnyObject(
+      (revision as unknown) as RawPayload,
+      {
+        uuid: revision.item_uuid,
+      }
+    );
     const encryptedPayload = CreateSourcedPayloadFromObject(
       payload,
       PayloadSource.RemoteHistory
     );
-    const decryptedPayload = await this.protocolService!.payloadByDecryptingPayload(
+    const decryptedPayload = await this.protocolService.payloadByDecryptingPayload(
       encryptedPayload
     );
-    return new ItemHistoryEntry(decryptedPayload);
+    if (decryptedPayload.errorDecrypting) {
+      return undefined;
+    }
+    return new HistoryEntry(decryptedPayload as SurePayload);
+  }
+
+  optimizeHistoryForItem(uuid: string): void {
+    /**
+     * Clean up if there are too many revisions. Note itemRevisionThreshold
+     * is the amount of revisions which above, call for an optimization. An
+     * optimization may not remove entries above this threshold. It will
+     * determine what it should keep and what it shouldn't. So, it is possible
+     * to have a threshold of 60 but have 600 entries, if the item history deems
+     * those worth keeping.
+     *
+     * Rules:
+     * - Keep an entry if it is the oldest entry
+     * - Keep an entry if it is the latest entry
+     * - Keep an entry if it is Significant
+     * - If an entry is Significant and it is a deletion change, keep the entry before this entry.
+     */
+    const entries = this.history[uuid] || [];
+    if (entries.length <= this.itemRevisionThreshold) {
+      return;
+    }
+
+    const isEntrySignificant = (entry: HistoryEntry) => {
+      return entry.deltaSize() > LargeEntryDeltaThreshold;
+    };
+    const keepEntries: HistoryEntry[] = [];
+    const processEntry = (
+      entry: HistoryEntry,
+      index: number,
+      keep: boolean
+    ) => {
+      /**
+       * Entries may be processed retrospectively, meaning it can be
+       * decided to be deleted, then an upcoming processing can change that.
+       */
+      if (keep) {
+        keepEntries.unshift(entry);
+        if (isEntrySignificant(entry) && entry.operationVector() === -1) {
+          /** This is a large negative change. Hang on to the previous entry. */
+          const previousEntry = entries[index + 1];
+          if (previousEntry) {
+            keepEntries.unshift(previousEntry);
+          }
+        }
+      } else {
+        /** Don't keep, remove if in keep */
+        removeFromArray(keepEntries, entry);
+      }
+    };
+    for (let index = entries.length; index--; ) {
+      const entry = entries[index];
+      if (index === 0 || index === entries.length - 1) {
+        /** Keep the first and last */
+        processEntry(entry, index, true);
+      } else {
+        processEntry(entry, index, isEntrySignificant(entry));
```

Comment:
You could do
```ts
const isEntrySignificant = index === 0 || index === entries.length || isEntrySignificant(entry);
```
And remove the need for a nested function altogether to do your processing inline, thus having a clean up-to-down reading flow